### PR TITLE
[CALCITE-3924] Fix flakey test to handle TIMESTAMP and TIMESTAMP(0) correctly

### PR DESCRIPTION
**Pull request by neoremind**
https://issues.apache.org/jira/browse/CALCITE-3924

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java [hunk 1]
```diff
@@ -180,8 +180,10 @@ protected void generateTypeString(StringBuilder sb, boolean withDetail) {
     boolean printScale = scale != SCALE_NOT_SPECIFIED;
 
     // for the digest, print the precision when defaulted,
-    // since (for instance) TIME is equivalent to TIME(0).
-    if (withDetail) {
+    // but for TIME or TIMESTAMP, although (for instance)
+    // TIME is equivalent to TIME(0), the digest of TIME
+    // and TIME(0) should be different literally.
+    if (withDetail && !SqlTypeName.DATETIME_TYPES.contains(typeName)) {
       // -1 means there is no default value for precision
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:187]: Use `SqlTypeUtil.isDatetime`

Comment (neoremind) [reply core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:187]: `SqlTypeUtil.isDatetime` accepts `RelDataType`, but here the type is `SqlTypeName`, so I was thinking it is not able to change. 
`RelDataType --contains--> SqlTypeName --contains--> SqlTypeFamily`

Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:187]: The whole if block should be removed. Not only the `DATETIME_TYPES ` has this problem, all the types that allows precision could have such a problem, the root cause is that `#toString` and `digest` has different handling of printing non specified precision.

One way to solve is to not fire caching of types without precision and with default precision.

Comment (neoremind) [reply core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java:187]: The code is very old, back to the first commit in 2012, considering minimum change, I add this check. Disabling `canonize` for some types with conditions would also add more logic. So I am thinking that removing the whole condition block (all unit tests will pass) would be a good solution.

Comment (neoremind) [reply core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java]: comment addressed.

**Pull request closed without merging**